esp/services/esdl_svc_engine/esdl_monitor.cpp

```diff
@@ -0,0 +1,576 @@
+                {
+                    CEsdlInstance* instance = item.getValue();
+                    if (!instance || !instance->m_def.get())
+                        continue;
+                    if (!instance->m_def->isShared())
+                        ids.append(instance->m_defId.str());
+                }
+                ForEachItemIn (x, ids)
+                {
+                    const char* id = ids.item(x);
+                    DBGLOG("Definition %s is no longer being used, remove it from esdl shared cache", id);
+                    m_esdlInstanceMap.remove(id);
+                }
+            }
+        }
+
+        return 0;
+    }
+
+    virtual void stop()
+    {
+        m_stop = true;
+        m_waitsem.signal();
+    }
+private:
+    CriticalSection m_critSect;
+    MapStringTo<Owned<CEsdlInstance>> m_esdlInstanceMap;
+    Semaphore m_waitsem;
+    bool m_stop;
+};
+
+static CriticalSection gEsdlMonitorCritSection;
+
+class CEsdlMonitor : implements IEsdlMonitor, public CInterface, implements IEsdlListener
+{
+private:
+    MapStringTo<EsdlBindingImpl*> m_esdlBindingMap;
+    CriticalSection m_CritSect;
+    StringBuffer mEnvptHeader;
+    Owned<IEsdlStore> m_pCentralStore;
+    Owned<IEsdlSubscription> m_pSubscription;
+    Owned<CEsdlShare> m_esdlShare;
+    bool m_isSubscribed;
+
+public:
+    IMPLEMENT_IINTERFACE;
+
+    CEsdlMonitor() : m_isSubscribed(false)
+    {
+        constructEnvptHeader();
+        m_pCentralStore.setown(createEsdlCentralStore());
+        m_esdlShare.setown(new CEsdlShare());
+        m_esdlShare->start();
+        DBGLOG("EsdlMonitor started.");
+    }
+
+    virtual ~CEsdlMonitor()
+    {
+        if (m_pSubscription)
+            m_pSubscription->unsubscribe();
+    }
+
+    CEsdlShare* queryEsdlShare()
+    {
+        return m_esdlShare.get();
+    }
+
+    void setupSubscription()
+    {
+        m_isSubscribed = true;
+        m_pSubscription.setown(createEsdlSubscription(this));
+    }
+
+    virtual void subscribe() override
+    {
+        CriticalBlock cb(m_CritSect);
+        if(m_isSubscribed)
+            return;
+        m_isSubscribed = true;
+        m_pSubscription->subscribe();
+    }
+
+    virtual void unsubscribe() override
+    {
+        CriticalBlock cb(m_CritSect);
+        if(!m_isSubscribed)
+            return;
+        m_isSubscribed = false;
+        m_pSubscription->unsubscribe();
+    }
+
+    //Reference count increment is done by the function
+    virtual void registerBinding(const char* bindingId, IEspRpcBinding* binding) override
+    {
+        if (!bindingId || !binding)
+            return;
+        EsdlBindingImpl* esdlbinding = dynamic_cast<EsdlBindingImpl*>(binding);
+        if (esdlbinding)
+        {
+            CriticalBlock cb(m_CritSect);
+            m_esdlBindingMap.setValue(bindingId, esdlbinding);
+        }
+    }
+
+    void loadDynamicBindings()
+    {
+        IPropertyTree * esdlBindings = m_pCentralStore->getBindings();
```

Comment:
getXXX normally implies it returns a linked item, so it looks like this is going to leak..